executor/runtime/docker/docker.go

```diff
@@ -1640,6 +1649,137 @@ func (r *DockerRuntime) Start(parentCtx context.Context, pod *v1.Pod) (string, *
 	return logDir, details, statusMessageChan, err
 }
 
+func (r *DockerRuntime) reportContainerStatusMetrics(ctx context.Context) {
+	l := log.WithField("taskID", r.c.TaskID())
+	defer func() {
+		if r := recover(); r != nil {
+			l.Warnf("Unexpected panic when reporting container status metrics: %#v", r)
+		}
+	}()
+	spectatordSocket := fmt.Sprintf("/var/lib/titus-inits/%s/root/run/spectatord/spectatord.unix", r.c.TaskID())
+	conn, err := connectToSpectatord(spectatordSocket, 3) // Retry 3 times before giving up.
+	if err != nil {
+		l.WithError(err).Warnf("Dial error with spectatord socket %s", spectatordSocket)
+		return
+	}
+	defer conn.Close()
+	reportTicker := time.NewTicker(1 * time.Minute)
+	defer reportTicker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			l.WithError(ctx.Err()).Debug("Stopping reporting container status metric")
+			return
+		case <-reportTicker.C:
+			metricLines := r.containerStatusMetrics()
+			for _, m := range metricLines {
+				if _, err := conn.Write([]byte(m)); err != nil {
+					l.WithError(err).Warnf("Error sending metric to spectatord, metric string: %s", m)
+				}
+			}
+		}
+	}
+}
+
+func connectToSpectatord(spectatordSocket string, retriesLeft int) (net.Conn, error) {
+	conn, err := net.Dial("unixgram", spectatordSocket)
+	if err != nil {
+		if retriesLeft--; retriesLeft > 0 {
+			time.Sleep(1 * time.Second)
+			return connectToSpectatord(spectatordSocket, retriesLeft)
+		}
+	}
+	return conn, err
+}
+
+func (r *DockerRuntime) containerStatusMetrics() []string {
+	l := log.WithField("taskID", r.c.TaskID())
+	var metricLines []string
+	// 1. Loop through all containers.
+	for _, c := range r.c.Pod().Status.ContainerStatuses {
+		// 2. Get container metadata and state.
+		metricTags := make(map[string]string)
+		metricTags["nf.container"] = c.Name
+		metricTags["nf.process"] = processNameForContainer(c.Name, r.c.Pod())
```

Comment:
@copperlight  should this simply not be set if we don't have a PROCESS_NAME?
As is this code would set it to the empty string.

Reply:
Empty tag values are filtered out during sanitization: https://github.com/Netflix/titus-executor/blob/TITUS-5663/executor/runtime/docker/docker.go#L1771.

Reply:
Correct, do not add the tag if it has no value. If that's removed during a sanitization step, then that should be fine.